### DISCUSSION TITLE
Form validation not working

### DISCUSSION THREAD

**Discussion by cristianhosu** [element-plus/element-plus#1758] [Q&A]
Hello,

I have the following code which fails to validate properly:

```typescript
<template>
	<el-form ref="form" :model="productData" :rules="formRules" label-position="left" label-width="150px" class="form">
		<div class="column">
			<el-form-item label="Code" prop="code">
				<el-input v-model="productData.code" placeholder="Code" />
			</el-form-item>
			<el-form-item label="Name" prop="name">
				<el-input v-model="productData.name" placeholder="Name" />
			</el-form-item>
		</div>
	</el-form>
</template>
<script lang="ts">
import { defineComponent, onMounted, reactive, ref } from 'vue';

export default defineComponent({
	setup(props: any) {
		const required = (field: string, callback: any) => {
			if ((productData.value as any)[field]) callback();
			else {
				callback(new Error('Field is mandatory'));
			}
		};

		const form = ref();
		const productData = ref({});
		const formRules = reactive({
			code: [
				{
					validator: (rule: any, value: any, callback: any) => {
						required('code', callback);
					},
					trigger: 'blur',
				},
			],
			// code: [{ required: true, message: 'Field is mandatory', trigger: 'blur' }],
			name: [
				{
					validator: (rule: any, value: any, callback: any) => {
						required('name', callback);
					},
					trigger: 'blur',
				},
			],
		});

		return {
			form,
			productData,
			formRules
		};
	},
});
</script>
```

Inside the custom validator function, value is null. If i check the productData.value[field] there is a value when i write something in the input.

(I have another form done in a similar manner which is working. The only difference between them is that this one is in a component which is loaded in another component under an el-tab-pane, the one that is working is loaded in a el-dialog)

What am I missing?

Any hints, highly appreciated.

Thank you

**Comment by mcrio** [comment]
I just took a brief look at the code so I might be wrong here:

- You don't need reactive form rules. Those won't change.
- `productData` is not of a primitive type, so should be instantiated using `reactive`.
- The validator can just return `Error` or `Error[]` (Elemen Plus does not render all errors by default)  or `true` on success.

I'm assuming the problem you are describing is tied to point 2 from above.

**Reply by jw-foss** [reply]
- Yeah, rules should remain non-reactive for performance.
- Both `ref` and `reactive` is acceptable here. When using `ref`, the `ref` object will be automatically unwrapped into `reactive`, it is basically `reactive` after all.
- Passing both `Error` and `Error[]` are acceptable I suppose. Passing `undefined` or `true` means the validation holds true.

I dunno what is really going on with this case, I made a dry run case which seemed ok.

**Comment by jw-foss** [comment]
Sorry for the late response, maybe next time you should open an issue instead of discussion since issues are our high priority in this project.
Dry running the code worked for me, see: https://codesandbox.io/s/form-validation-0i64f?file=/src/App.vue
